cedar-policy/src/api/err.rs

```diff
@@ -121,65 +326,63 @@ pub enum SchemaError {
-    #[error("unsupported feature used in schema: {0}")]
-    UnsupportedFeature(String),
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    UnsupportedFeature(UnsupportedFeatureError),
     /// Undeclared entity type(s) used in the `memberOf` field of an entity
     /// type, the `appliesTo` fields of an action, or an attribute type in a
     /// context or entity attribute record. Entity types in the error message
     /// are fully qualified, including any implicit or explicit namespaces.
-    #[error("undeclared entity type(s): {0:?}")]
-    UndeclaredEntityTypes(HashSet<String>),
-    /// Undeclared action(s) used in the `memberOf` field of an action.
-    #[error("undeclared action(s): {0:?}")]
-    UndeclaredActions(HashSet<String>),
-    /// Undeclared common type(s) used in entity or context attributes.
-    #[error("undeclared common type(s): {0:?}")]
-    UndeclaredCommonTypes(HashSet<String>),
+    /// Or undeclared action(s) used in the `memberOf` field of an action.
+    /// Or undeclared common type(s) used in entity or context attributes.
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    UndeclaredSchemaConstruct(UndeclaredSchemaConstructError),
     /// Duplicate specifications for an entity type. Argument is the name of
     /// the duplicate entity type.
-    #[error("duplicate entity type `{0}`")]
-    DuplicateEntityType(String),
-    /// Duplicate specifications for an action. Argument is the name of the
+    /// Or duplicate specifications for an action. Argument is the name of the
     /// duplicate action.
-    #[error("duplicate action `{0}`")]
-    DuplicateAction(String),
-    /// Duplicate specification for a reusable type declaration.
-    #[error("duplicate common type `{0}`")]
-    DuplicateCommonType(String),
+    /// Or duplicate specification for a reusable type declaration.
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    DuplicateSchemaConstruct(DuplicateSchemaConstructError),
     /// Cycle in the schema's action hierarchy.
-    #[error("cycle in action hierarchy containing `{0}`")]
-    CycleInActionHierarchy(EntityUid),
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    CycleInActionHierarchy(CycleInActionHierarchyError),
     /// Cycle in the schema's common type declarations.
-    #[error("cycle in common type references containing `{0}`")]
-    CycleInCommonTypeReferences(Name),
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    CycleInCommonTypeReferences(CycleInCommonTypeReferencesError),
     /// The schema file included an entity type `Action` in the entity type
     /// list. The `Action` entity type is always implicitly declared, and it
     /// cannot currently have attributes or be in any groups, so there is no
     /// purposes in adding an explicit entry.
     #[error("entity type `Action` declared in `entityTypes` list")]
     ActionEntityTypeDeclared,
```

Comment:
In think these unit enum variants should also contain an empty struct so that we can add details inside the struct in the future